Remotion/Data/DomainObjects/Persistence/ReadOnlyStorageProviderDecorator.cs

```diff
@@ -0,0 +1,121 @@
+using Remotion.Data.DomainObjects.Queries.Configuration;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence
+{
+  /// <summary>
+  /// Decorator for <see cref="IReadOnlyStorageProvider"/> to ensure that <see cref="IQuery"/>s have the appropriate <see cref="QueryType"/> when
+  /// being executed in the ReadOnly context.
+  /// </summary>
+  public class ReadOnlyStorageProviderDecorator : IReadOnlyStorageProvider
+  {
+    public IReadOnlyStorageProvider InnerStorageProvider { get; }
+
+    public ReadOnlyStorageProviderDecorator (IReadOnlyStorageProvider innerStorageProvider)
+    {
+      ArgumentUtility.CheckNotNull("innerStorageProvider", innerStorageProvider);
+
+      InnerStorageProvider = innerStorageProvider;
+    }
+
+    public void Dispose ()
+    {
+      InnerStorageProvider.Dispose();
+    }
+
+    public ObjectLookupResult<DataContainer> LoadDataContainer (ObjectID id)
+    {
+      ArgumentUtility.CheckNotNull("id", id);
+
+      return InnerStorageProvider.LoadDataContainer(id);
+    }
+
+    public IEnumerable<ObjectLookupResult<DataContainer>> LoadDataContainers (IReadOnlyCollection<ObjectID> ids)
+    {
+      ArgumentUtility.CheckNotNull("ids", ids);
+
+      return InnerStorageProvider.LoadDataContainers(ids);
+    }
+
+    public IEnumerable<DataContainer> LoadDataContainersByRelatedID (
+        RelationEndPointDefinition relationEndPointDefinition,
+        SortExpressionDefinition? sortExpressionDefinition,
+        ObjectID relatedID)
+    {
+      ArgumentUtility.CheckNotNull("relationEndPointDefinition", relationEndPointDefinition);
+      ArgumentUtility.CheckNotNull("relatedID", relatedID);
+
+      return InnerStorageProvider.LoadDataContainersByRelatedID(relationEndPointDefinition, sortExpressionDefinition, relatedID);
+    }
+
+    public IEnumerable<DataContainer?> ExecuteCollectionQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      CheckQueryType(query);
+
+      return InnerStorageProvider.ExecuteCollectionQuery(query);
+    }
+
+    public IEnumerable<IQueryResultRow> ExecuteCustomQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      CheckQueryType(query);
+
+      return InnerStorageProvider.ExecuteCustomQuery(query);
+    }
+
+    public object? ExecuteScalarQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      CheckQueryType(query);
+
+      return InnerStorageProvider.ExecuteScalarQuery(query);
+    }
+
+    public void BeginTransaction ()
+    {
+      InnerStorageProvider.BeginTransaction();
+    }
+
+    public void Commit ()
+    {
+      InnerStorageProvider.Commit();
+    }
+
+    public void Rollback ()
+    {
+      InnerStorageProvider.Rollback();
+    }
+
+    private void CheckQueryType (IQuery query)
+    {
+      if (query.QueryType is QueryType.CollectionReadWrite or QueryType.CustomReadWrite or QueryType.ScalarReadWrite)
```

Comment:
I forgot while writing the initial review comment:

Do you think it has added value to check for allowed instead of forbidden QueryType values?

With allowed values, we make sure to list only those we actually support.

With denied values, if there's a new value introduced that's also incorrect, the argument check would miss it.

I'm not sure if it's worth it though at this point, but in my opionion it's the safer and thus cleaner approach. If you think the argument is compelling enough to actually update the code, then please do. 

@martin-herwich  might also have an opinion on it.